File: src/csharp/Intel/Iced.UnitTests/Intel/FormatterTests/FormatterTestCases.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if GAS || INTEL || MASM || NASM || FAST_FMT
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Iced.Intel;

namespace Iced.UnitTests.Intel.FormatterTests {
	public readonly struct FormatterTestCase {
		public readonly int Bitness;
		public readonly string HexBytes;
		public readonly ulong IP;
		public readonly Code Code;
		public readonly DecoderOptions Options;
		public FormatterTestCase(int bitness, string hexBytes, ulong ip, Code code, DecoderOptions options) {
			Bitness = bitness;
			HexBytes = hexBytes;
			IP = ip;
			Code = code;
			Options = options;
		}
	}

	static class FormatterTestCases {
		static (FormatterTestCase[] tests, HashSet<int> ignored) tests16;
		static (FormatterTestCase[] tests, HashSet<int> ignored) tests32;
		static (FormatterTestCase[] tests, HashSet<int> ignored) tests64;
		static (FormatterTestCase[] tests, HashSet<int> ignored) tests16_Misc;
		static (FormatterTestCase[] tests, HashSet<int> ignored) tests32_Misc;
		static (FormatterTestCase[] tests, HashSet<int> ignored) tests64_Misc;

		public static (FormatterTestCase[] tests, HashSet<int> ignored) GetTests(int bitness, bool isMisc) {
			if (isMisc) {
				return bitness switch {
					16 => GetTests(ref tests16_Misc, bitness, isMisc),
					32 => GetTests(ref tests32_Misc, bitness, isMisc),
					64 => GetTests(ref tests64_Misc, bitness, isMisc),
					_ => throw new ArgumentOutOfRangeException(nameof(bitness)),
				};
			}
			else {
				return bitness switch {
					16 => GetTests(ref tests16, bitness, isMisc),
					32 => GetTests(ref tests32, bitness, isMisc),
					64 => GetTests(ref tests64, bitness, isMisc),
					_ => throw new ArgumentOutOfRangeException(nameof(bitness)),
				};
			}
		}

		static (FormatterTestCase[] tests, HashSet<int> ignored) GetTests(ref (FormatterTestCase[] tests, HashSet<int> ignored) data, int bitness, bool isMisc) {
			if (data.tests is null) {
				var filename = "InstructionInfos" + bitness;
				if (isMisc)
					filename += "_Misc";
				data.ignored = new HashSet<int>();
				data.tests = GetTests(filename, bitness, data.ignored).ToArray();
			}
			return data;
		}

		static readonly char[] sep = new[] { ',' };
		static IEnumerable<FormatterTestCase> GetTests(string filename, int bitness, HashSet<int> ignored) {
			int lineNo = 0;
			int testCaseNo = 0;
			filename = FileUtils.GetFormatterFilename(filename);
			foreach (var line in File.ReadLines(filename)) {
				lineNo++;
				if (line.Length == 0 || line[0] == '#')
					continue;
				var parts = line.Split(sep);
				DecoderOptions options;
				if (parts.Length == 2)
					options = default;
				else if (parts.Length == 3) {
					if (!ToEnumConverter.TryDecoderOptions(parts[2].Trim(), out options))
						throw new InvalidOperationException($"Invalid line #{lineNo} in file {filename}");
				}
				else
					throw new InvalidOperationException($"Invalid line #{lineNo} in file {filename}");
				var hexBytes = parts[0].Trim();
				var codeStr = parts[1].Trim();
				var ip = bitness switch {
					16 => DecoderConstants.DEFAULT_IP16,
					32 => DecoderConstants.DEFAULT_IP32,
					64 => DecoderConstants.DEFAULT_IP64,
					_ => throw new InvalidOperationException(),
				};
				if (CodeUtils.IsIgnored(codeStr))
					ignored.Add(testCaseNo);
				else {
					if (!ToEnumConverter.TryCode(codeStr, out var code))
						throw new InvalidOperationException($"Invalid line #{lineNo} in file {filename}");
					yield return new FormatterTestCase(bitness, hexBytes, ip, code, options);
				}
				testCaseNo++;
			}
		}

		public static IEnumerable<object[]> GetFormatData(int bitness, string formatterDir, string formattedStringsFile, bool isMisc = false) {
			var data = GetTests(bitness, isMisc);
			var formattedStrings = FileUtils.ReadRawStrings(Path.Combine(formatterDir, $"Test{bitness}_{formattedStringsFile}")).ToArray();
			return GetFormatData(data.tests, data.ignored, formattedStrings);
		}

		static IEnumerable<object[]> GetFormatData(FormatterTestCase[] tests, HashSet<int> ignored, string[] formattedStrings) {
			formattedStrings = Utils.Filter(formattedStrings, ignored);
			if (tests.Length != formattedStrings.Length)
				throw new ArgumentException($"(tests.Length) {tests.Length} != (formattedStrings.Length) {formattedStrings.Length} . tests[0].HexBytes = {(tests.Length == 0 ? "<EMPTY>" : tests[0].HexBytes)} & formattedStrings[0] = {(formattedStrings.Length == 0 ? "<EMPTY>" : formattedStrings[0])}");
			var res = new object[tests.Length][];
			for (int i = 0; i < tests.Length; i++)
				res[i] = new object[3] { i, tests[i], formattedStrings[i] };
			return res;
		}

		public static IEnumerable<object[]> GetFormatData(int bitness, (string hexBytes, Instruction instruction)[] tests, string formatterDir, string formattedStringsFile) {
			var formattedStrings = FileUtils.ReadRawStrings(Path.Combine(formatterDir, $"Test{bitness}_{formattedStringsFile}")).ToArray();
			return GetFormatData(tests, formattedStrings);
		}

		static IEnumerable<object[]> GetFormatData((string hexBytes, Instruction instruction)[] tests, string[] formattedStrings) {
			if (tests.Length != formattedStrings.Length)
				throw new ArgumentException($"(tests.Length) {tests.Length} != (formattedStrings.Length) {formattedStrings.Length} . tests[0].hexBytes = {(tests.Length == 0 ? "<EMPTY>" : tests[0].hexBytes)} & formattedStrings[0] = {(formattedStrings.Length == 0 ? "<EMPTY>" : formattedStrings[0])}");
			var res = new object[tests.Length][];
			for (int i = 0; i < tests.Length; i++)
				res[i] = new object[3] { i, tests[i].instruction, formattedStrings[i] };
			return res;
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/fmt/FormatterTestCases.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt;

import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashSet;

import org.junit.jupiter.params.provider.*;

import com.github.icedland.iced.x86.CodeUtils;
import com.github.icedland.iced.x86.DecoderConstants;
import com.github.icedland.iced.x86.FileUtils;
import com.github.icedland.iced.x86.ToCode;
import com.github.icedland.iced.x86.ToDecoderOptions;
import com.github.icedland.iced.x86.dec.DecoderOptions;
import com.github.icedland.iced.x86.dec.NonDecodedTestCase;

public final class FormatterTestCases {
	public static class Info {
		public FormatterTestCase[] tests;
		public HashSet<Integer> ignored;
	}

	static final Info tests16;
	static final Info tests32;
	static final Info tests64;
	static final Info tests16_Misc;
	static final Info tests32_Misc;
	static final Info tests64_Misc;

	static {
		tests16 = createTests(16, false);
		tests32 = createTests(32, false);
		tests64 = createTests(64, false);
		tests16_Misc = createTests(16, true);
		tests32_Misc = createTests(32, true);
		tests64_Misc = createTests(64, true);
	}

	public static Info getTests(int bitness, boolean isMisc) {
		if (isMisc) {
			switch (bitness) {
			case 16:
				return tests16_Misc;
			case 32:
				return tests32_Misc;
			case 64:
				return tests64_Misc;
			default:
				throw new UnsupportedOperationException();
			}
		}
		else {
			switch (bitness) {
			case 16:
				return tests16;
			case 32:
				return tests32;
			case 64:
				return tests64;
			default:
				throw new UnsupportedOperationException();
			}
		}
	}

	private static Info createTests(int bitness, boolean isMisc) {
		String filename = "InstructionInfos" + bitness;
		if (isMisc)
			filename += "_Misc";
		Info info = new Info();
		info.ignored = new HashSet<Integer>();
		info.tests = readTests(filename, bitness, info.ignored).toArray(new FormatterTestCase[0]);
		return info;
	}

	private static ArrayList<FormatterTestCase> readTests(String filename, int bitness, HashSet<Integer> ignored) {
		int lineNo = 0;
		int testCaseNo = 0;
		filename = FmtFileUtils.getFormatterFilename(filename);
		ArrayList<String> lines = FileUtils.readAllLines(filename);
		ArrayList<FormatterTestCase> result = new ArrayList<FormatterTestCase>();
		for (String line : lines) {
			lineNo++;
			if (line.length() == 0 || line.charAt(0) == '#')
				continue;
			String[] parts = line.split(",");
			int options;
			if (parts.length == 2)
				options = DecoderOptions.NONE;
			else if (parts.length == 3) {
				Integer value = ToDecoderOptions.tryGet(parts[2].trim());
				if (value == null)
					throw new UnsupportedOperationException(String.format("Invalid line #%d in file %s", lineNo, filename));
				options = value;
			}
			else
				throw new UnsupportedOperationException(String.format("Invalid line #%d in file %s", lineNo, filename));
			String hexBytes = parts[0].trim();
			String codeStr = parts[1].trim();
			long ip;
			switch (bitness) {
			case 16:
				ip = DecoderConstants.DEFAULT_IP16;
				break;
			case 32:
				ip = DecoderConstants.DEFAULT_IP32;
				break;
			case 64:
				ip = DecoderConstants.DEFAULT_IP64;
				break;
			default:
				throw new UnsupportedOperationException();
			}
			if (CodeUtils.isIgnored(codeStr))
				ignored.add(testCaseNo);
			else {
				Integer value = ToCode.tryGet(codeStr);
				if (value == null)
					throw new UnsupportedOperationException(String.format("Invalid line #%d in file %s", lineNo, filename));
				int code = value;
				result.add(new FormatterTestCase(bitness, hexBytes, ip, code, options));
			}
			testCaseNo++;
		}
		return result;
	}

	public static Iterable<Arguments> getFormatData(int bitness, String formatterDir, String formattedStringsFile) {
		return getFormatData(bitness, formatterDir, formattedStringsFile, false);
	}

	public static Iterable<Arguments> getFormatData(int bitness, String formatterDir, String formattedStringsFile, boolean isMisc) {
		Info info = getTests(bitness, isMisc);
		String[] formattedStrings = FmtFileUtils
				.readRawStrings(Paths.get(formatterDir, String.format("Test%d_%s", bitness, formattedStringsFile)).toString()).toArray(new String[0]);
		return getFormatData(info.tests, info.ignored, formattedStrings);
	}

	private static Iterable<Arguments> getFormatData(FormatterTestCase[] tests, HashSet<Integer> ignored, String[] formattedStrings) {
		formattedStrings = Utils.filter(formattedStrings, ignored);
		if (tests.length != formattedStrings.length)
			throw new UnsupportedOperationException();
		ArrayList<Arguments> result = new ArrayList<Arguments>(tests.length);
		for (int i = 0; i < tests.length; i++)
			result.add(Arguments.of(i, tests[i], formattedStrings[i]));
		return result;
	}

	public static Iterable<Arguments> getFormatData(int bitness, NonDecodedTestCase[] tests, String formatterDir, String formattedStringsFile) {
		String[] formattedStrings = FmtFileUtils
				.readRawStrings(Paths.get(formatterDir, String.format("Test%d_%s", bitness, formattedStringsFile)).toString()).toArray(new String[0]);
		return getFormatData(tests, formattedStrings);
	}

	private static Iterable<Arguments> getFormatData(NonDecodedTestCase[] tests, String[] formattedStrings) {
		if (tests.length != formattedStrings.length)
			throw new UnsupportedOperationException();
		ArrayList<Arguments> result = new ArrayList<Arguments>(tests.length);
		for (int i = 0; i < tests.length; i++)
			result.add(Arguments.of(i, tests[i].instruction, formattedStrings[i]));
		return result;
	}
}
